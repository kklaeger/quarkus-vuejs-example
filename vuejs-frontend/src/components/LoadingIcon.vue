<template>
  <b-loading :is-full-page="false" v-model="isLoading"></b-loading>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
@Component({
  components: {}
})
export default class LoadingIcon extends Vue {
  private isLoading = true;
}
</script>

<style scoped lang="scss"></style>
