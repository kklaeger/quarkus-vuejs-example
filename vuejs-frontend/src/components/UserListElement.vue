<template>
  <div>
    <div class="columns is-mobile">
      <div class="column is-3-mobile is-2-tablet is-1-desktop is-2-widescreen">
        <b-icon pack="fas" icon="user" size="is-medium"> </b-icon>
      </div>
      <div class="column name-column">{{ user.name }}</div>
      <div class="column is-3-mobile is-2-tablet is-1-desktop is-2-widescreen">
        <b-button rounded type="is-text" @click="deleteUser">
          <b-icon pack="fas" icon="trash" size="is-medium"> </b-icon>
        </b-button>
      </div>
      <div v-if="isLoading" class="loading">
        <LoadingIcon />
      </div>
    </div>
    <div v-if="hasError" class="error">
      Unable to delete user: {{ errorMsg }}
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from "vue-property-decorator";
import { User } from "@/models/User";
import LoadingIcon from "@/components/LoadingIcon.vue";

@Component({
  components: { LoadingIcon }
})
export default class UserListElement extends Vue {
  @Prop({ required: true }) user!: User;

  get isLoading() {
    return (this.$store.state.deleteUserIsLoading as string[]).includes(
      this.user.id
    );
  }

  get hasError() {
    return (this.$store.state.deleteUserHasError as string[]).includes(
      this.user.id
    );
  }

  get errorMsg() {
    return (this.$store.state.deleteUserErrorMsg as { [id: string]: string })[
      this.user.id
    ];
  }

  deleteUser(): void {
    this.$store.dispatch("deleteUser", this.user.id);
  }
}
</script>

<style scoped lang="scss">
@import "./../styles/variables.scss";
.name-column {
  text-align: left;
  padding-top: 15px;
  padding-bottom: 15px;
}

.error {
  padding-bottom: 15px;
  color: $error-message;
}
</style>
