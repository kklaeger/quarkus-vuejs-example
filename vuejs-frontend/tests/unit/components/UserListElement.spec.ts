import { shallowMount, createLocalVue, mount } from "@vue/test-utils";
import Vuex from "vuex";
import Buefy from "buefy";
import UserListElement from "@/components/UserListElement.vue";
import { User } from "@/models/User";
import LoadingIcon from "@/components/LoadingIcon.vue";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Buefy);

describe(UserListElement.name + " component", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render successfully.", () => {
    const user: User = { id: "1", name: "Test" };
    const state = {
      users: [user],
      fetchUsersIsLoading: false,
      fetchUsersHasError: false,
      addUserIsLoading: false,
      addUserHasError: false,
      deleteUserIsLoading: [],
      deleteUserHasError: [],
      deleteUserErrorMsg: {}
    };
    const actions = {
      deleteUser: jest.fn()
    };
    const store = new Vuex.Store({
      state,
      actions
    });

    const wrapper = shallowMount(UserListElement, {
      store,
      localVue,
      propsData: { user }
    });

    expect(wrapper.text()).toMatch("Test");
    expect(wrapper).toMatchSnapshot();
  });

  it("should click delete button and call respective dispatch.", async () => {
    const user: User = { id: "1", name: "user1" };
    const state = {
      users: [user],
      fetchUsersIsLoading: false,
      fetchUsersHasError: false,
      addUserIsLoading: false,
      addUserHasError: false,
      deleteUserIsLoading: [],
      deleteUserHasError: [],
      deleteUserErrorMsg: {}
    };
    const actions = {
      deleteUser: jest.fn()
    };
    const store = new Vuex.Store({
      state,
      actions
    });

    const wrapper = mount(UserListElement, {
      store,
      localVue,
      propsData: { user }
    });

    await wrapper.findComponent({ name: "BButton" }).trigger("click");

    expect(actions.deleteUser.mock.calls[0][1]).toEqual("1");
    expect(wrapper.findComponent({ name: "BButton" }).exists()).toBeTruthy();
    expect(wrapper.find(".error").exists()).toBeFalsy();
    expect(wrapper.find(".loading").exists()).toBeFalsy();
    expect(wrapper.findComponent(LoadingIcon).exists()).toBeFalsy();
    expect(wrapper).toMatchSnapshot();
  });

  it("should display loading icon if id is in deleteUserIsLoading.", async () => {
    const user: User = { id: "1", name: "user1" };
    const state = {
      users: [user],
      fetchUsersIsLoading: false,
      fetchUsersHasError: false,
      addUserIsLoading: false,
      addUserHasError: false,
      deleteUserIsLoading: ["1"],
      deleteUserHasError: [],
      deleteUserErrorMsg: {}
    };
    const actions = {
      deleteUser: jest.fn()
    };
    const store = new Vuex.Store({
      state,
      actions
    });

    const wrapper = shallowMount(UserListElement, {
      store,
      localVue,
      propsData: { user }
    });

    expect(wrapper.find(".error").exists()).toBeFalsy();
    expect(wrapper.find(".loading").exists()).toBeTruthy();
    expect(wrapper.findComponent(LoadingIcon).exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it("should display error icon if id is in deleteUserHasError.", async () => {
    const user: User = { id: "1", name: "user1" };
    const state = {
      users: [user],
      fetchUsersIsLoading: false,
      fetchUsersHasError: false,
      addUserIsLoading: false,
      addUserHasError: false,
      deleteUserIsLoading: [],
      deleteUserHasError: ["1"],
      deleteUserErrorMsg: { "1": "xyz" }
    };
    const actions = {
      deleteUser: jest.fn()
    };
    const store = new Vuex.Store({
      state,
      actions
    });

    const wrapper = shallowMount(UserListElement, {
      store,
      localVue,
      propsData: { user }
    });

    expect(wrapper.find(".error").exists()).toBeTruthy();
    expect(wrapper.find(".error").text()).toEqual("Unable to delete user: xyz");
    expect(wrapper.find(".loading").exists()).toBeFalsy();
    expect(wrapper.findComponent(LoadingIcon).exists()).toBeFalsy();
    expect(wrapper).toMatchSnapshot();
  });
});
