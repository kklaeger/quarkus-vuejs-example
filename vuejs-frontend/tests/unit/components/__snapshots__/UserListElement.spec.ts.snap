// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserListElement component should click delete button and call respective dispatch. 1`] = `
<div>
  <div class="columns is-mobile">
    <div class="column is-3-mobile is-2-tablet is-1-desktop is-2-widescreen"><span class="icon is-medium"><i class="fas fa-user fa-lg"></i></span></div>
    <div class="column name-column">user1</div>
    <div class="column is-3-mobile is-2-tablet is-1-desktop is-2-widescreen"><button type="button" class="button is-text is-rounded">
        <!----><span><span class="icon is-medium"><i class="fas fa-trash fa-lg"></i></span></span>
        <!---->
      </button></div>
    <!---->
  </div>
  <!---->
</div>
`;

exports[`UserListElement component should display error icon if id is in deleteUserHasError. 1`] = `
<div>
  <div class="columns is-mobile">
    <div class="column is-3-mobile is-2-tablet is-1-desktop is-2-widescreen">
      <b-icon-stub pack="fas" icon="user" size="is-medium"></b-icon-stub>
    </div>
    <div class="column name-column">user1</div>
    <div class="column is-3-mobile is-2-tablet is-1-desktop is-2-widescreen">
      <b-button-stub type="is-text" rounded="true" nativetype="button" tag="button">
        <b-icon-stub pack="fas" icon="trash" size="is-medium"></b-icon-stub>
      </b-button-stub>
    </div>
    <!---->
  </div>
  <div class="error">
    Unable to delete user: xyz
  </div>
</div>
`;

exports[`UserListElement component should display loading icon if id is in deleteUserIsLoading. 1`] = `
<div>
  <div class="columns is-mobile">
    <div class="column is-3-mobile is-2-tablet is-1-desktop is-2-widescreen">
      <b-icon-stub pack="fas" icon="user" size="is-medium"></b-icon-stub>
    </div>
    <div class="column name-column">user1</div>
    <div class="column is-3-mobile is-2-tablet is-1-desktop is-2-widescreen">
      <b-button-stub type="is-text" rounded="true" nativetype="button" tag="button">
        <b-icon-stub pack="fas" icon="trash" size="is-medium"></b-icon-stub>
      </b-button-stub>
    </div>
    <div class="loading">
      <loadingicon-stub></loadingicon-stub>
    </div>
  </div>
  <!---->
</div>
`;

exports[`UserListElement component should render successfully. 1`] = `
<div>
  <div class="columns is-mobile">
    <div class="column is-3-mobile is-2-tablet is-1-desktop is-2-widescreen">
      <b-icon-stub pack="fas" icon="user" size="is-medium"></b-icon-stub>
    </div>
    <div class="column name-column">Test</div>
    <div class="column is-3-mobile is-2-tablet is-1-desktop is-2-widescreen">
      <b-button-stub type="is-text" rounded="true" nativetype="button" tag="button">
        <b-icon-stub pack="fas" icon="trash" size="is-medium"></b-icon-stub>
      </b-button-stub>
    </div>
    <!---->
  </div>
  <!---->
</div>
`;
